File: 2024/day3/day3.exs
defmodule Day3 do
  @pathname "input.txt"
  @pattern ~r/(do\(\))|(don't\(\))|(mul\((\d+),(\d+)\))/

  def run do
    File.read!(@pathname)
    |> scan()
    |> process()
    |> IO.inspect()
  end

  def scan(string) do
    Regex.scan(@pattern, string)
  end

  def process(_list, _capture \\ true, _sum \\ 0)
  def process([["do()" | _] | tail], _capture, sum), do: process(tail, true, sum)
  def process([["don't()" | _] | tail], _capture, sum), do: process(tail, false, sum)

  def process([[_, _, _, _, a, b] | tail], true, sum) do
    sum = sum + String.to_integer(a) * String.to_integer(b)
    process(tail, true, sum)
  end

  def process([_ | tail], false, sum), do: process(tail, false, sum)

  def process([], _capture, sum), do: sum
end

Day3.run()

File: 2024/day3/day3.py
import re

line = open("input.txt").read()

mul_instructions = re.findall(r"mul\((\d+),(\d+)\)", line)

# Pt1
print("Part one:", sum([int(a) * int(b) for a, b in mul_instructions]))


# Pt2
should_apply = True
sum = 0
for match in re.findall(r"(don't\(\))|(do\(\))|mul\((\d+),(\d+)\)", line):
    dont, do, a, b = match
    if dont:
        should_apply = False
    elif do:
        should_apply = True
    else:
        if should_apply:
            sum += int(a) * int(b)

print("Part 2:", sum)
